### PR TITLE
DAOS-14317 vos: initial changes for the phase2 object pre-load

### PR DESCRIPTION
**Pull request by NiuYawei**
- Introduced new durable format 'vos_obj_p2_df' for the md-on-ssd phase2
  object, at most 4 evict-able bucket IDs could be stored.

- Changed vos_obj_hold() & vos_obj_release() to pin or unpin object
  respectively.

- Changed the private data of VOS dkey/akey/value trees from 'vos_pool' to
  'vos_object', the private data will be used for allocating/reserving from
  the evict-able bucket.

- Move the vos_obj_hold() call from vos_update_end() to vos_update_begin()
  for the phase2 pool, reserve value from the object evict-able bucket.

TODO:
- Reorg GC code to pre-load objects before starting transaction.
- Reorg DTX commit code to pre-load objects before starting transaction.
- Reorg multiple-objects tx code to pre-load objects.
- Improve engine scheduler to take both non-evict-able & evict-able zones
  space pressure into account.

Required-githooks: true

Signed-off-by: Niu Yawei <yawei.niu@intel.com>

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate owners.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

File: src/tests/ftest/util/telemetry_utils.py
Comment (daltonbohning) [inline src/tests/ftest/util/telemetry_utils.py:427]: What does 360 vs 384 represent?

Comment (NiuYawei) [reply src/tests/ftest/util/telemetry_utils.py:427]: that's the size of 'vos_object' (a in-memory data structure), I don't know why the author put the size in the metric name though.

**Comment by NiuYawei** [comment]
The test failure is likely caused by DAOS-16465, I've pushed a fix to master, will merge the fix once it's landed.

**Comment by NiuYawei** [comment]
CI failures are from 2 known issues: DAOS-16367 & DAOS-16374

**Review by wangshilong — APPROVED** [review]

**Pull request merged**
